src/client-side-encryption/clientEncryption.ts

```diff
@@ -0,0 +1,1015 @@
+    this._keyVaultClient = options.keyVaultClient || client;
+    const MongoCrypt = ClientEncryption.getMongoCrypt();
+    this._mongoCrypt = new MongoCrypt(mongoCryptOptions);
+  }
+
+  /**
+   * Creates a data key used for explicit encryption and inserts it into the key vault namespace
+   *
+   * @example
+   * ```ts
+   * // Using callbacks to create a local key
+   * clientEncryption.createDataKey('local', (err, dataKey) => {
+   *   if (err) {
+   *     // This means creating the key failed.
+   *   } else {
+   *     // key creation succeeded
+   *   }
+   * });
+   * ```
+   *
+   * @example
+   * ```ts
+   * // Using async/await to create a local key
+   * const dataKeyId = await clientEncryption.createDataKey('local');
+   * ```
+   *
+   * @example
+   * ```ts
+   * // Using async/await to create an aws key
+   * const dataKeyId = await clientEncryption.createDataKey('aws', {
+   *   masterKey: {
+   *     region: 'us-east-1',
+   *     key: 'xxxxxxxxxxxxxx' // CMK ARN here
+   *   }
+   * });
+   * ```
+   *
+   * @example
+   * ```ts
+   * // Using async/await to create an aws key with a keyAltName
+   * const dataKeyId = await clientEncryption.createDataKey('aws', {
+   *   masterKey: {
+   *     region: 'us-east-1',
+   *     key: 'xxxxxxxxxxxxxx' // CMK ARN here
+   *   },
+   *   keyAltNames: [ 'mySpecialKey' ]
+   * });
+   * ```
+   */
+  createDataKey(
+    provider: KMSProvider,
+    options?: ClientEncryptionCreateDataKeyProviderOptions,
+    callback?: Callback<DataKey>
+  ) {
+    if (typeof options === 'function') {
+      callback = options;
+      options = {};
+    }
+    if (options == null) {
+      options = {};
+    }
+
+    const dataKey = Object.assign({ provider }, options.masterKey);
+
+    if (options.keyAltNames && !Array.isArray(options.keyAltNames)) {
+      throw new TypeError(
+        `Option "keyAltNames" must be an array of strings, but was of type ${typeof options.keyAltNames}.`
+      );
+    }
+
+    let keyAltNames = undefined;
+    if (options.keyAltNames && options.keyAltNames.length > 0) {
+      keyAltNames = options.keyAltNames.map((keyAltName, i) => {
+        if (typeof keyAltName !== 'string') {
+          throw new TypeError(
+            `Option "keyAltNames" must be an array of strings, but item at index ${i} was of type ${typeof keyAltName}`
+          );
+        }
+
+        return serialize({ keyAltName });
+      });
+    }
+
+    let keyMaterial = undefined;
+    if (options.keyMaterial) {
+      keyMaterial = serialize({ keyMaterial: options.keyMaterial });
+    }
+
+    const dataKeyBson = serialize(dataKey);
+    const context = this._mongoCrypt.makeDataKeyContext(dataKeyBson, {
+      keyAltNames,
+      keyMaterial
+    });
+    const stateMachine = new StateMachine({
+      proxyOptions: this._proxyOptions,
+      tlsOptions: this._tlsOptions
+    });
+
+    // @ts-expect-error We did not convert promiseOrCallback to TS
```

Comment:
We will remove callback support once this PR is merged - I intentionally did not TSify `promiseOrCallback` or `maybePromise`.